### PR TITLE
Validate and fix fuzzy records in `HollowAccountFinalizationSuite`

### LINKED ISSUE #9696: Fuzzy match records with mono-service for HollowAccountFinalizationSuite
- [x]   hollowAccountCompletionWithCryptoTransfer(),
- [x]                 hollowAccountCompletionWithContractCreate(),
- [x]                 hollowAccountCompletionWithContractCall(),
- [x]                 hollowAccountCompletionWithTokenAssociation(),
- [x]                 hollowAccountCompletionWithTokenTransfer(),
- [x]                 hollowAccountCompletionViaNonReqSigIsNotAllowed(),
- [x]                 hollowAccountCompletionWhenHollowAccountSigRequiredInOtherReqSigs(),
- [x]                 tooManyHollowAccountFinalizationsShouldFail(),
- [x]                 completedHollowAccountsTransfer(),
- [x]                 hollowAccountFinalizationWhenAccountNotPresentInPreHandle(),
- [x]                 hollowAccountFinalizationOccursOnlyOnceWhenMultipleFinalizationTensComeInAtTheSameTime(),
- [x]                 txnWith2CompletionsAndAnother2PrecedingChildRecords(),
- [x]                 hollowPayerAndOtherReqSignerBothGetCompletedInASingleTransaction(),
- [x]                 hollowAccountCompletionIsPersistedEvenIfTxnFails(),
- [x]                 precompileTransferFromHollowAccountWithNeededSigFailsAndDoesNotFinalizeAccount()

### PR DESCRIPTION
**Pull request by Neeharika-Sompalli**
Fixes #9696 

**Description**
 - Adds `CallOutcome#recipientId` to match mono-service logic for setting the called `ContractID` in the receipt of even a `ContractCall` or `EthereumTransaction`.
 - Adds [HandleException#shouldRollbackStack](https://github.com/hashgraph/hedera-services/blob/c9715afed4d79c74ed23f4924d7262ff590c8c34/hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/HandleException.java#L39) to let contract operations suppress the `HandleWorkflow` rollback their gas charges. (_Note:_ Once we rationalize contract fee charging this can be removed.)
 - Updates [CryptoUpdateRecordBuilder.accountId ](https://github.com/hashgraph/hedera-services/blob/c9715afed4d79c74ed23f4924d7262ff590c8c34/hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java#L567)only for hollow account finalizations
 - Skips verifying keys for a hollow payer
 - Adds optionalKeys set to preHandleResult. Changed [HandleWorkflow#haveKeyChanges](https://github.com/hashgraph/hedera-services/blob/c9715afed4d79c74ed23f4924d7262ff590c8c34/hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java#L848) to looks at the `preHandleResult.requiredKeys()`  and `preHandleResult.optionalKeys()`
 - Adds `calculateFees` for ContractCreateHandler
 - Ensures [royalty fallback fee is charged only onc](https://github.com/hashgraph/hedera-services/blob/c9715afed4d79c74ed23f4924d7262ff590c8c34/hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/customfees/CustomRoyaltyFeeAssessor.java#L94)e per tokenId

### CODE REVIEW

**Review by tinker-michaelj — DISMISSED** [review]
LGTM, ty @Neeharika-Sompalli !

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java [hunk 1]
```diff
@@ -684,11 +691,22 @@ private ValidationResult validate(
...
+     */
     private void rollback(
+            final boolean rollbackStack,
             @NonNull final ResponseCodeEnum status,
             @NonNull final SavepointStackImpl stack,
             @NonNull final RecordListBuilder recordListBuilder) {
-        stack.rollbackFullStack();
+        if (rollbackStack) {
+            stack.rollbackFullStack();
+        }
```
Comment (mhess-swl) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:709]: Using the stack rollback mechanism for everything _but_ gas feels weird. Is there really not a way we can charge gas fees without having to pass a boolean? E.g. some point in the handle workflow, or a common entrypoint for smart contracts

**Review by tinker-michaelj — DISMISSED** [review]
LGTM

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java (hunk 1 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:709]: Yes we will fix that post-migration to mod-service.

Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:709]: It seems all changes look like balance changes in account store for gas charged. So, no other way to differentiate as I understood

Comment (mhess-swl) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java:709]: Ok cool. A `// FUTURE` reminder would be nice (but not required unless tests are already failing 😅)

**Review by mhess-swl — APPROVED** [review]

**Review by iwsimon — APPROVED** [review]
Lots of changes! LGTM.

**Review by tinker-michaelj — APPROVED** [review]
LGTM!

**Pull request merged**